### ISSUE TITLE
Mine & Slash 1.12.2 Deep Dive Discussion [ACTIVE]

### ISSUE THREAD

**Issue by Paladin1979** [kreezxil/World-of-Dragons#335]
# Mod Nominated for Removal
Mod Name: Mine and Slash

# Reason for Nomination
Reason:
Just to be clear: I love the concept of WoD as described under "Pack Direction" on Twitch, and because Mine and Slash largely covers that concept, it seems like a good fit. I just think there are too much issues with the current implementation for it to work. Below is my reasoning (which comes down to my preference NOT to use the mod; the direction the mod is taking just doesn't work well enough for me.)

There are two things that I really do not like about M&S, which are the following:
- It does NOT work with a very basic idea in minecraft: that you can create your own stuff. Vanilla weapons are rendered useless. The current "fix" we have where vanilla weapons are automatically "randomized" at the players level after crafting breaks it the other way - no point in using M&S weaponry anymore (After leveling, a very effective method is just to craft 10 stone swords, and select the best one. The durability will be low but everything else is awesome, so that's OK. It's also a really easy way of obtaining M&S ores through salvaging). This may be "fixable" by limiting crafted items to "common" stats, and scaling the weapon damage based on the vanilla damage, but that doesn't feel great either.
- The energy bar effectively replaces the "default" Attack Speed mechanism in minecraft by something that is, IMHO, no better. And personally I'm of the opinion: If you can't improve it, don't change it. I would prefer it if the M&S weapons just have an Attack Speed like the vanilla weapons.

Stuff I like about M&S:
- Giving everything levels seems to work fine and is much in line with the idea behind WoD as I understand it.
- I like that the regular, expendable "enchantment" levels are completely separate - probably the only right way to implement an RPG-like level progression.
- I like the Diablo/WoW-style randomized loot

Something I used to like but what seems to have changed:
- Before the last update of WoD 18-something, it looked like areas had levels; now, everything just spawns at the player level. If everything always spawns at my own level, what's the point? The area-levels have their own problems (doesn't work well with /rtp) but I do not think the current solution works. In the concept, I'd think it works best if players somehow have to move to progressively higher areas. The /rtp could maybe be fixed by creating a new area at that player's current level (with some areas 5-10 levels higher around it) - fine for building a base, but you can't stick around long.

Stuff that can be improved (tweaking some numbers):
- The amount of experience needed to level, the way health of mobs scales, etc., needs a lot of fine-tuning. I tested some of the XP progression between level 1 and 10 and there's really weird stuff there.
- Level difficulty is essentially linear, which means that going from level 1 to 2 is a MUCH bigger deal than going from 10 to 11 (or even 10 to 15); that doesn't feel right to me. Basically - levels mean less as you advance further. (in contrast, in WoW, anything which is 5 levels higher than you is always really tough, (almost) everything 5 levels below is easy - use exponential advancement rather than linear)
- How much time is it supposed to take to get to level 80? If I set up a mob grinder and kill them as fast as I can (feels like cheating), it still takes a LONG time to level up. There is no real other way. Is that really what you want? Can be fixed by fixing the numbers, but this needs thought.

Stuff I would like to see:
- MUCH better integration with the vanilla crafting and enchantment system (and also any weapons, armor and baubles added by modpacks). It should be balanced such that crafting your own gear is a great way to get started or a backup if you get really unlucky drops - but the randomized loot should yield something better within a reasonable amount of time. Crafting/enchanting "good" stuff should be possible but should take much more resources.

**Comment by hellwolf95** [comment]
Well paladin your first statement is rendered useless due to the fact default weapons are not added to the compatiable list of m&s weapons. Same with modded weapons and armors.

We cant tweak mob hp and damage scaling.

Also on a second note building mob farms on krees server will make me go find it amd reset those chunks or ban you. We have warned players against it many times.


Although i do agree certain mods need to be made compatiable and the player stats need to be tweaked.
I dont agree with the stamina complaint its very easy to get used to meaning you arnt trying to get used to it.

And any further complaints about m&s take to the m&s team as we dont have control over lots of those systems.

**Comment by Paladin1979** [comment]
- Can you clarify why exactly that renders the statement useless? Just because the behaviour is intended, does not mean I have to like it. This is my main argument for removing the mod from the pack: bad synergy with (almost) everything else. The mod authors seem to like it this way, I don't think it works well or fits in with WoD.

- If it is not possible to tweak hp or damage scaling, that would (to me) be all the more reason to remove the mod: it means that (in my opinion) it cannot be customized enough.

- "Also on a second note building mob farms on krees server will make me go find it amd reset those chunks or ban you. We have warned players against it many times."
To clarify: I made the mob farm in SP (single player) to test some of M&S aspects. I wanted to get a more solid view on how the mod works before I suggested removing it. I have not made, and _do not_ intend to make a mob farm on the server, as I am well aware that this not allowed for various reasons. I can see how you misinterpreted this, but please do not make threats about banning me based on stuff I did not intend, or do.
If it is not possible to scale the xp progression, again, all the more reason to remove it from the pack.

- The stamina complaint does _not_ mean I haven't tried getting used to it. I have, and I am used to it. I just don't _like_ it. It's a minor point, but please do not draw conclusions based on what I did not write.

This is _not_ a complaint about M&S; these are reasons why I think it should be removed from the pack. Bad synergy with (most) other mods and an inability to finetune it being the most important ones.

**Comment by hellwolf95** [comment]
What i mean is that kree can just add the weapons to the compat list and they will have levels. He just doesnt want to do that

**Comment by kreezxil** [comment]
I'm changing the title of this to please make it more compat, because I want to leave it open as the discussion seems sane enough to continue

# however i will not remove this mod

**Comment by Paladin1979** [comment]
I agree with Kreezxil on that point; just adding them as compatible doesn't really work. As I stated - my current best way to obtain a decent sword at my own level is not to grind mobs, but just to craft a bunch of stone swords. The random enchants mean some of them will be awesome (I got lvl 6 stone swords ranging from 25 to 50 damage per hit, I believe), AND I can salvage the rest to repair the first.

(oh, and combining 2 low-level damaged swords creates an undamaged one at your own level...)

Theorizing how it might be improved:

It might work better if (for instance) the mod would modify a weapon's base damage rather than rerolling it, and never generate anything but common quality items from crafting. It would require the basic drops of M&S to be much better versions of the vanilla weapons. If everything is based on modified vanilla weapons, it would be much easier to apply the same to weapons/armor from other packs.

You would be able to craft basic weapons at your own level - but you would have to fight mobs to get _good_ weapons. Which you would really need to fight the boss fights, or the occasional legendary spawn of a regular mob.

If repairing items was based on regular ores/items (iron/gold/diamond nuggets, bits of stone, shards of obsidian, etc) it might be easier to do, but would integrate better. Add occasional "special" salvage drops that you can use for cool stuff (upgrading weapons, etc.).

**Comment by hellwolf95** [comment]
What you are suggesting is changes to the mod itself. Kree does not own that mod so any major changes like that need to be directed at the mod dev

**Comment by AzureZhen** [comment]
I am the maintainer of the 1.12 version of Mine and Slash. I am will look into something for crafted weapons, most likely making them cap at a certain level/rarity. As for the leveling yes, 1 to 8 requires far less experience but after level 8, the needed amount does in fact increase. This is intended.

**Comment by Paladin1979** [comment]
Hi Azure, out of curiosity: is any of these suggested changes likely to resonate with the Mine & Slash team (I understand they focus on 1.14 now and have declared 1.12 obsolete) and/or is it more likely we would have to make our own changes on the 1.12 version?

**Comment by AzureZhen** [comment]
I handle anything 1.12 for Mine and Slash myself so ill see what I can do but I'm open to any suggestions or help. You can find my fork here https://github.com/AzureZhen/Mine-and-Slash

**Comment by Paladin1979** [comment]
Never done minecraft mods or java, but I do have a lot of software developing experience - I might take a look. Don't have much time, though, so don't expect anything anytime soon.

**Comment by Paladin1979** [comment]
I thought I should add a note about the XP progression, especially after the last update. The xp curve needed to level up is much more sane now; I also noticed that xp for mobs of higher level than yourself is "cut off" at the xp you would get for that mob at your own level (I again did some testing in single-player)

So: two thumbs up for porting the xp progression; this makes it much less of a bother to get those first 20 levels before adventure maps start dropping, which is a big deal.

There's still a couple of things that I would change though, if possible - based on my experience of the past week. Not sure if any of this is something in 1.14 that can be ported to 1.12, though. I'm perfectly happy to copy-paste this to the Mine and Slash people as well.
- I would not let animals give mine and slash xp or loot. Currently, a lot of my xp comes from slaughtering cows and squid, and actually some of my better gear too. This feels very wrong and exploitable. (currently animals give half the xp of hostile mobs) Another option I thought of would be to make the xp reward always based on a level 1 mob, which means slaughtering animals would help getting past the first few levels, but would hardly be worth it after (not in terms of xp, anyway). I would drop the loot in any case.
- Instead of penalizing low-level players for making high-level kills (the xp cutoff I just mentioned), I would do it the other way around, i.e. don't give a player xp for killing a mob 5(?) levels lower than him/herself. Diablo and WoW have quite sophisticated scaling systems for the xp mobs give to players of lower and higher levels that can be used as examples.
- If a player kills a mob of a higher level than himself (more power to him if he can manage it), change the loot dropped to the level of the player (I think this is what Diablo does). I found it very frustrating that, although I was very much capable of killing level 20 mobs, that I had to go back to a level 10 area to get gear I could use. There is something nice about having a legendary level 23 sword waiting for me, but the frustration of not having usable gear now is worse.
- Any options for bridging the gap until adventure maps start dropping at level 20 would be welcome. This is the main reason I set up my base in a level 20 area - I actually CAN use the adventure maps that drop every once in a while. Why is this limitation even there (i.e. why not drop level 1 adventure maps?)
- There's an exploit where a summoned zombie from electroblob's gives full xp to the player who killed it. It would be better if summoned mobs yield no xp at all, but instead have the summoner yield more. This could actually be a solution for camping mob spawners, where breaking the spawner if what yields mine and slash xp (equal to 4 times a mob or something), rather than killing the mobs that come out of it. The spawner itself might then need to get a rarity though. Complicated.
- having the FTB quests reward mine and slash xp and/or items (swords, armor) might be a good way of encouraging/helping players through the first 20 levels, and be a big incentive to follow the quest book.

**Comment by jam325** [comment]
Paladin that are some good ideas for the mod, but kreezxil can't make anything about it you need to go to the mod author with these ideas so someting can change but kree has no might about the mine & slash mod

**Comment by AzureZhen** [comment]
@Paladin1979 
- the issue with animals right now is that its either give animals levels and such or not be able to use CustomNPCs with the mod as they share the same base code sadly. It is configurable though already.
- hmm I'll see what I can do.
- the issue is the loot dropped is always generator at the mobs level, this is a system I actually like and won't be changing as it's working as intended.
- this is configurable, the reason is due to the cost of the map device and i dont to want to overpack a newbies inventories with useless maps.
- yeah the mob summons is something I'll have to blacklist in the exp gain.
- I've already suggested this in another issue. There are givexp and givegear commands that FTBQ can run as OP to give players as quests rewards.

@jam325 luckily I'm already here.

**Comment by kreezxil** [comment]
Please don't forget you guys have discussion running here.

**Comment by Paladin1979** [comment]
OK, so after having worked with the updated system for a while: now that the new XP progression is in place, leveling goes a lot faster - which makes both the issue of maps not dropping until level 20 and higher level gear dropping less of an issue.

As a side note, I've raised a couple of these issues in the mine and slash github as well, but I think it is safe to say that most of my ideas for balancing/improving the mod are diametrically opposed to what the mod author intends for it - so that ends there. I'll limit myself to suggesting changes to how the mod is tailored within this pack.

I'd still suggest lowering the point where maps drop to level 10 or so - even though the map device is expensive (which would be fixable by adjusting the recipe to something cheaper - I honestly think both the map device and the salvage station are way too expensive - leveling involves killing a lot of mobs, and the spawn rate is a lot better on the adventure maps.

On the WoD 18 server, I actually had a plan to build a couple of one-chunk "waystations" at level 1, 5, 10 and 15 areas, with a gear repair and salvage station inside each and connected by nether portals, to give starting players a leg up with the mod - protecting them in such a way that everyone could use them but they couldn't be griefed.

This was precisely because the salvage station feels (at least to me) central to the mod - it gives you a way to get rid of all those excess drops with bad stats and get something useful in return. Without it, you either have to stash all that junk for later recycling (takes up a lot of space) or trash it (which feels wasteful) - but the station is just too expensive to make early on. Same for the map device and gear modify station. (the modify station allows you to use runes, which start dropping at level 3 - long before you can actually build the station (unless you're a lucky diamond miner - probably))

The gear crafting station is a different matter, I don't have a problem with that not being available until later. TBH that thing feels completely useless, but that's a whole different story (which mostly touches on my dislike that the mod has bad compatibility with crafting items, which is a central theme to minecraft and most mods)

The mod author also seemed to suggest it is possible to configure the rarity level of compatible items, e.g. making it so that a stone sword gets maximum "uncommon" stats or something. No idea how that would work, there's no documentation I can find - I hope Azure knows more about it.

So summarizing:

- I suggest lowering the level where maps start to drop to 10 or so
- I suggest changing the crafting recipes of the salvage station, map device and modify station to be a lot cheaper
- I suggest re-configuring the compatibility of vanilla items (and perhaps those of other mods) to be maximum "uncommon" (if it can be finetuned that far, it could be material dependent, i.e. a wooden sword would be max "common", a stone sword max "uncommon", an iron sword max "rare" and no limits on diamond. Or something like that.
- it might be cool to have pre-generated structures spawn in the world with a salvage and repair station inside (maybe in villages?) - this could be an alternative to making the recipes cheaper

**Comment by AzureZhen** [comment]
1. Thats at the pack level as the config needs updated.
2. I agree on this and is on my todo list.
3. That's easy to do.
4. I've never messed with adding to Villages, I'll see what I can do.

**Comment by AzureZhen** [comment]
Per the re-configuring of vanilla items. This is what I have planned so far. 
**Weapons:**
Wooden = Max level of 10, always normal (never runed), max common, not able to salvaged.
Stone = Max level of 20, always normal (never runed), max uncommon, not able to salvaged.
Iron = Max level of 30, always normal (never runed), max rare, not able to salvaged.
Gold = Max level of 30, always normal (never runed), max rare, not able to salvaged.
Diamond = Minimum level of 50, not able to salvaged.
Bow = Minimum level of 5, not able to salvaged.

**Armor:**
Leather = Max level of 10, always normal (never runed), max common, not able to salvaged.
Iron = Max level of 30, always normal (never runed), max rare, not able to salvaged.
Golden = Max level of 30, always normal (never runed), max rare, not able to salvaged.
Diamond = Minimum level of 50, not able to salvaged.
Chain Mail = Minimum level of 40, able to be salvaged since it can't be crafted.

**Comment by AzureZhen** [comment]
![image](https://user-images.githubusercontent.com/7415711/67255505-632d8780-f450-11e9-88e1-bc15809be7c6.png)
![image](https://user-images.githubusercontent.com/7415711/67255509-658fe180-f450-11e9-897a-f2edc27ed162.png)
![image](https://user-images.githubusercontent.com/7415711/67255512-67f23b80-f450-11e9-8229-9d10997c9af9.png)
Redid the recipes

**Comment by kreezxil** [comment]
map drops in todays patch starts at level 10 and one or another the cheaper
recipes will come in to play. I do have crafttweaker.

**Comment by Paladin1979** [comment]
nice! I would argue for making diamond items salvageable, since diamonds are not a renewable resource (to my best knowledge). Although you're probably better off pulling them through the pulverizer to regain some diamonds.

**Comment by kreezxil** [comment]
you can dig for diamonds, don't be afraid to go mining.

**Comment by AzureZhen** [comment]
You should be able to pulverize, i just made it so they aren't salvageable for the M&S Salvage Station.
